---
title: "Using R lang client for koina"
author:
- name: Ludwig Lautenbacher
  affiliation:
    - Computational Mass Spectrometry, Technical University of Munich (TUM), Freising, Germany
  email: ludwig.lautenbacher@tum.de
- name: Christian Panse
  affiliation:
    - Functional Genomics Center Zurich (FGCZ) - University of Zurich | ETH Zurich, Winterthurerstrasse 190, CH-8057 Zurich, Switzerland
    - Swiss Institute of Bioinformatics (SIB), Quartier Sorge - Batiment Amphipole, CH-1015 Lausanne, Switzerland
package: koinar
abstract: |
  How to use `r BiocStyle::Biocpkg('KoinaR')` to fetch predictions from [Koina](https://koina.wilhelmlab.org/)
output:
  BiocStyle::html_document:
    toc_float: true
bibliography: koina.bib
vignette: >
  %\usepackage[utf8]{inputenc}
  %\VignetteIndexEntry{On using the R lang client for koina}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
urlcolor: blue
---

```{r style, results = 'asis'}
BiocStyle::markdown()
knitr::opts_chunk$set(fig.wide = TRUE, fig.retina = 3, error=FALSE, eval=TRUE)
```

# Introduction
Koina is a community driven endeavor to make AI models for life science research accessible. It is a service that tries to reduce the barrier of entry for using machine learning models in life science research by removing the need to set up and maintain the infrastructure required to run the models. The project is open-source and adheres to FAIR principles.

Here we take a look at KoinaR the R package to simplify interfacing with Koina. 

## Basic usage

```{r create, eval=TRUE}
# Create a client tied to a specific server & model
# Here we use the model published by Gessulat et al [@prosit2019]
# And the public server available at koina.wilhelmlab.org
# All available models can be found at https://koina.wilhelmlab.org/docs
prosit2019 <- koinar::Koina$new(
  model_name = "Prosit_2019_intensity",
  server_url = "koina.wilhelmlab.org:443"
)
prosit2019
```



```{r input}
# Create example inputs
# Here we look at two different peptide sequences with charge 1 and 2 respectively
input <- data.frame(
  peptide_sequences = c("LGGNEQVTR", "GAGSSEPVTGLDAK"),
  collision_energies = c(25, 25),
  precursor_charges = c(1, 2)
)

# Fetch the predictions by calling `$predict` of the model you want to use
# A progress bar shows you how much of the predictions are already done
# In this case this should complete instantly
prediction_results <- prosit2019$predict(input)

# Display the predictions
# The output varies depending on the chosen model
# For the intenstiy model we get a data.frame with 5 columns 
# The three inputs we provided: peptide_sequences, collision_energies, precursor_charges 
# and for each predicted fragment ion: annotation, mz, intensities
head(prediction_results)
```


## Example 1: Reproducing Fig.1d from [@prosit2019]

```{r Fig1, echo=FALSE, out.width="100%", eval=TRUE, fig.cap="Screenshot of Fig.1d [@prosit2019] https://www.nature.com/articles/s41592-019-0426-7"}
  knitr::include_graphics("https://user-images.githubusercontent.com/4901987/237583247-a948ceb3-b525-4c30-b701-218346a30cf6.jpg")
```

```{r Fig2, out.width="100%", fig.cap="LKEATIQLDELNQK CE35 3+; Reproducing Fig1", fig.height=8, fig.retina=3}
prosit2019

## Use the peptide_sequence collision_energy and
## precursor_charge as in the figure
input <- data.frame(
 peptide_sequences = c("LKEATIQLDELNQK"),
 collision_energies = c(35),
 precursor_charges = c(3)
)

prediction_results <- prosit2019$predict(input)

# Plot the spectrum
plot(prediction_results$intensities ~ prediction_results$mz, type = 'n',
     ylim = c(0, 1.1))
yIdx <- grepl("y", prediction_results$annotation)
points(prediction_results$mz[yIdx], prediction_results$intensities[yIdx],
       col = "red", type = 'h', lwd = 2)
points(prediction_results$mz[!yIdx], prediction_results$intensities[!yIdx],
       col = "blue", type = 'h', lwd = 2)

text(prediction_results$mz, prediction_results$intensities,
     labels = prediction_results$annotation,
     las = 2, cex = 1, pos =3)

```

## Example 2: Biognosys iRT fragment ions prediction

```{r defineBiognosysIrtPeptides}
iRTpeptides <- c("LGGNEQVTR", "YILAGVENSK", "GTFIIDPGGVIR", "GTFIIDPAAVIR",
                 "GAGSSEPVTGLDAK", "TPVISGGPYEYR", "VEATFGVDESNAK",
                 "TPVITGAPYEYR", "DGLDAASYYAPVR", "ADVTPADFSEWSK",
                 "LFLQFGAQGSPFLK")

input <- data.frame(
 peptide_sequences = iRTpeptides,
 collision_energies = 35,
 precursor_charges = 2
)
```

```{r AlphaPept}
prosit2019$predict((input)) -> rv.prosit2019
rv.prosit2019$model <- "Prosit_2019_intensity"

AlphaPept <- koinar::Koina$new(
  model_name = "AlphaPept_ms2_generic",
  server_url = "koina.wilhelmlab.org:443"
)
input$instrument_types = "QE"
AlphaPept$predict(input) -> rv.AlphaPept
rv.AlphaPept$model <- "AlphaPept_ms2_generic"
```

```{r xyplot, out.width="100%", fig.cap="iRT peptides fragment ins prediction using  AlphaPept and Prosit2019", fig.height=15, fig.retina=3}
lattice::xyplot(intensities ~ mz | model * peptide_sequences ,
                group = grepl('y', annotation),
                data = rbind(rv.AlphaPept[,names(rv.prosit2019)],
                             rv.prosit2019),
                type = 'h')
```

## Example 3: Spectrum similarity plot with OrgMassSpecR
```{r mssimplot, out.width="100%", fig.cap="Spectral similarity ms2pip vs prosit plot created with OrgMassSpecR"}
library(OrgMassSpecR)

input <- data.frame(
  peptide_sequences = c("LGGNEQVTR"),
  collision_energies = c(25),
  precursor_charges = c(2)
)

prosit2019 <- koinar::Koina$new("Prosit_2019_intensity")
ms2pip <- koinar::Koina$new("ms2pip_2021_HCD")

pred_prosit <- prosit2019$predict(input)
pred_ms2pip <- ms2pip$predict(input)

sim <- SpectrumSimilarity(pred_prosit[,c("mz", "intensities")],
                          pred_ms2pip[,c("mz", "intensities")],
                          top.lab = "Prosit",
                          bottom.lab = "MS2PIP",
                          b = 25)
title(main = paste("Spectrum similarity", round(sim, 3)))
```

## Example 4: Loading rawdata with the Spectra package
```{r}
library(Spectra)
library(msdata)
library(OrgMassSpecR)

# Provide the paths of rawfiles you wish to load.
# In this example we use data provided by the msdata package
fls <- c(
  system.file("proteomics", 
              "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz", 
              package = "msdata")
  )

# Load rawdata using the Spectra package
data <- Spectra(fls, source = MsBackendMzR())
# Filter rawdata for MS2 spectra only
data = filterMsLevel(data, msLevel = 2)
```

```{r}
# Extract metadata
metadata = spectraData(data)

# Extract spectra data
spectra = peaksData(data)
```

```{r}
# Sort data by precursor mass since mascot-queries are sorted by mass
metadata$mass = (metadata$precursorMz*metadata$precursorCharge)
peptide_mass_order = order(metadata$mass)
metadata = metadata[peptide_mass_order,]
sorted_spectra = spectra[peptide_mass_order]
```

```{r}
# Fetch prediction using KoinaR package
# The identifications for this data can be found at
# https://ftp.pride.ebi.ac.uk/pride/data/archive/2012/03/PXD000001/F063721.dat
# For this example we picked a random spectrum 4128
# Searching for q4128 at the above website we find the sequence 
# [UNIMOD:737]-AAVEEGVVAGGGVALIR
input <- data.frame(
  peptide_sequences = c("[UNIMOD:737]-AAVEEGVVAGGGVALIR"),
  collision_energies = c(45),
  precursor_charges = c(3),
  fragmentation_types = c("HCD")
)
prosit2019 <- koinar::Koina$new("Prosit_2020_intensity_TMT")
pred_prosit <- prosit2019$predict(input)
```

```{r}
# Plot spectral similarity between the experimental and predicted spectrum
# We can observe high agreement between the experimental and predicted spectrum.
# Validating this identification from Mascot.
sim <- SpectrumSimilarity(sorted_spectra[[4128]],
                          pred_prosit[,c("mz", "intensities")],
                          top.lab = "Experimental",
                          bottom.lab = "Prosit",
                          t=0.01)
title(main = paste("Spectrum similarity", round(sim, 3)))
```

# Install
```{r, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("koinar")
```

# References {-}

# Session information {-}
```{r sessioninfo, eval=TRUE}
sessionInfo()
```
